ykrt/src/trace/swt/mod.rs

```diff
@@ -0,0 +1,133 @@
+    static FUNC_NAMES: RefCell<HashMap<usize, CString>> = RefCell::new(HashMap::new());
+}
+
+extern "C" {
+    fn get_function_names(
+        section: *const BitcodeSection,
+        result: *mut *mut IRFunctionNameIndex,
+        len: *mut usize,
+    );
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct IRFunctionNameIndex {
+    pub index: usize,
+    pub name: *const libc::c_char,
+}
+/// Inserts LLVM IR basicblock metadata into a thread-local BASIC_BLOCKS vector.
+///
+/// # Arguments
+/// * `function_index` - The index of the function to which the basic block belongs.
+/// * `block_index` - The index of the basic block within the function.
+pub fn trace_basicblock(function_index: usize, block_index: usize) {
+    BASIC_BLOCKS.with(|v| {
+        v.borrow_mut().push(TracingBlock {
+            function_index,
+            block_index,
+        });
+    })
+}
+
+pub(crate) struct SWTracer {}
+
+impl super::Tracer for SWTracer {
+    fn start_recorder(self: Arc<Self>) -> Result<Box<dyn TraceRecorder>, Box<dyn Error>> {
+        BASIC_BLOCKS.with(|bbs| {
+            bbs.borrow_mut().clear();
+        });
+        Ok(Box::new(SWTTraceRecorder {}))
+    }
+}
+
+impl SWTracer {
+    pub fn new() -> Result<Self, Box<dyn Error>> {
+        Ok(SWTracer {})
+    }
+}
+
+struct SWTTraceRecorder {}
+
+impl TraceRecorder for SWTTraceRecorder {
+    fn stop(self: Box<Self>) -> Result<Box<dyn AOTTraceIterator>, InvalidTraceError> {
+        let mut aot_blocks: Vec<TraceAction> = vec![];
+        BASIC_BLOCKS.with(|tb| {
+            FUNC_NAMES.with(|fnames| {
+                FUNC_NAMES_INIT.call_once(|| {
+                    let mut functions: *mut IRFunctionNameIndex = std::ptr::null_mut();
+                    let bc_section = crate::compile::jitc_llvm::llvmbc_section();
+                    let mut functions_len: usize = 0;
+                    unsafe {
+                        get_function_names(
+                            &BitcodeSection {
+                                data: bc_section.as_ptr(),
+                                len: u64::try_from(bc_section.len()).unwrap(),
+                            },
+                            &mut functions,
+                            &mut functions_len,
+                        );
+                        for entry in std::slice::from_raw_parts(functions, functions_len) {
+                            fnames.borrow_mut().insert(
+                                entry.index,
+                                std::ffi::CStr::from_ptr(entry.name).to_owned(),
+                            );
+                        }
+                    }
+                });
+
+                aot_blocks = tb
+                    .borrow()
+                    .iter()
+                    .map(|tb| match fnames.borrow_mut().get(&tb.function_index) {
+                        Some(name) => TraceAction::MappedAOTBlock {
+                            func_name: name.to_owned(),
```

Comment:
@vext01 Looking at this makes me realise that having to pass around (owned) `String`s is really inefficient. Would it be possible to just carry around the address of the function or some other ID that can later be converted to a string?

[Obviously we'd need to do this with hwt too. And it wouldn't hold up this PR but it's a question worth thinking about generally.]